rust/cedar-policy-mcp-schema-generator/src/generator/request.rs

```diff
@@ -0,0 +1,2011 @@
+use mcp_tools_sdk::description::ServerDescription;
+use smol_str::{SmolStr, ToSmolStr};
+use uuid::Uuid;
+
+#[derive(Clone, Debug)]
+/// A `RequestGenerator` is a structure produced by a `SchemaGenerator` which allows
+/// a user to generator Cedar Requests that are valid for the Schema generated by the
+/// `SchemaGenerator`.
+pub struct RequestGenerator {
+    config: SchemaGeneratorConfig,
+    tools: ServerDescription,
+    root_namespace: Option<Name>,
+    schema: ValidatorSchema,
+}
+
+impl RequestGenerator {
+    pub(crate) fn new(
+        config: SchemaGeneratorConfig,
+        tools: ServerDescription,
+        root_namespace: Option<Name>,
+        schema: ValidatorSchema,
+    ) -> Self {
+        Self {
+            config,
+            tools,
+            root_namespace,
+            schema,
+        }
+    }
+
+    /// Generate an authorization request & entities necessary to call the Cedar authorization engine.
+    /// A user of this function is expected to provide
+    /// (1) the principal and resource of the request
+    /// (2) any common context elements
+    /// (3) any entity information for the principal, resource, and common context
+    /// (4) the MCP tool input request, and (5) optionally the MCP tool output response
+    ///
+    /// The function will then return a Cedar Request and entities necessary to determine if the principal is
+    /// authorized to use (or receive the oputout of) the requested tool.
+    pub fn generate_request(
+        &self,
+        principal: EntityUID,
+        resource: EntityUID,
+        context: impl IntoIterator<Item = (SmolStr, RestrictedExpr)>,
+        mut entities: Entities,
+        input: &Input,
+        output: Option<&Output>,
+    ) -> Result<(Request, Entities), RequestGeneratorError> {
+        let input = self.tools.validate_input(input)?;
+        // PANIC SAFETY: `self.tools` must contain a tool named `input.name()` and `input` validates against `tool`
+        #[allow(
+            clippy::unwrap_used,
+            reason = "Validation ensures there is a tool in with the same name that the input validates against"
+        )]
+        let tool = self
+            .tools
+            .tool_descriptions()
+            .find(|t| t.name() == input.name())
+            .unwrap();
+
+        let output = output
+            .map(|output| self.tools.validate_output(tool.name(), output))
+            .transpose()?;
+
+        let mut type_defs = self
+            .tools
+            .type_definitions()
+            .map(|ty_def| (ty_def.name().to_smolstr(), self.root_namespace.clone()))
+            .collect::<HashMap<_, _>>();
+
+        let tool_ns: Name = tool.name().parse()?;
+        let tool_ns = tool_ns.qualify_with_name(self.root_namespace.as_ref());
+
+        // collect all of the tool specific type defs
+        type_defs.extend(
+            tool.type_definitions()
+                .map(|ty_def| (ty_def.name().to_smolstr(), Some(tool_ns.clone()))),
+        );
```

Comment:
This `generate_request()` function is very long. You could consider factoring some functionality into helper functions. This highlighted portion could maybe become a helper function, or even its own struct `TypeDefs` with its own constructors and methods.